src/scene/shader-lib/chunks/lit/frag/lighting/lightEvaluation.js

```diff
@@ -0,0 +1,196 @@
+
+        // cookie attenuation
+        #if defined(LIGHT{i}COOKIE)
+
+            #if LIGHT{i}TYPE == SPOT
+                #ifdef LIGHT{i}COOKIE_FALLOFF
+                    #ifdef LIGHT{i}COOKIE_TRANSFORM
+                        dAtten3 = getCookie2DXform(light{i}_cookie, light{i}_shadowMatrix, light{i}_cookieIntensity, light{i}_cookieMatrix, light{i}_cookieOffset).{LIGHT{i}COOKIE_CHANNEL};
+                    #else
+                        dAtten3 = getCookie2D(light{i}_cookie, light{i}_shadowMatrix, light{i}_cookieIntensity).{LIGHT{i}COOKIE_CHANNEL};
+                    #endif
+                #else
+                    #ifdef LIGHT{i}COOKIE_TRANSFORM
+                        dAtten3 = getCookie2DClipXform(light{i}_cookie, light{i}_shadowMatrix, light{i}_cookieIntensity, light{i}_cookieMatrix, light{i}_cookieOffset).{LIGHT{i}COOKIE_CHANNEL};
+                    #else
+                        dAtten3 = getCookie2DClip(light{i}_cookie, light{i}_shadowMatrix, light{i}_cookieIntensity).{LIGHT{i}COOKIE_CHANNEL};
+                    #endif
+                #endif
+            #endif
+
+            #if LIGHT{i}TYPE == OMNI
+                dAtten3 = getCookieCube(light{i}_cookie, light{i}_shadowMatrix, light{i}_cookieIntensity).{LIGHT{i}COOKIE_CHANNEL};
+            #endif
+        #endif
+
+        // distance falloff
+        #if LIGHT{i}SHAPE == PUNCTUAL
+            #if LIGHT{i}FALLOFF == LINEAR
+                dAtten = getFalloffLinear(light{i}_radius, dLightDirW);
+            #else
+                dAtten = getFalloffInvSquared(light{i}_radius, dLightDirW);
+            #endif
+        #else
+            // non punctual lights only gets the range window here
+            dAtten = getFalloffWindow(light{i}_radius, dLightDirW);
+        #endif
+
+        if (dAtten > 0.00001) { // OPEN BRACKET
+
+            // spot light angle falloff
+            #if LIGHT{i}TYPE == SPOT
+                #if !defined(LIGHT{i}COOKIE) || defined(LIGHT{i}COOKIE_FALLOFF)
+                    dAtten *= getSpotEffect(light{i}_direction, light{i}_innerConeAngle, light{i}_outerConeAngle, dLightDirNormW);
+                #endif
+            #endif
+    #endif
+
+    // diffuse lighting - LTC lights do not mix diffuse lighting into attenuation that affects specular
+    #if LIGHT{i}SHAPE != PUNCTUAL
+        #if LIGHT{i}TYPE == DIRECTIONAL
+            // NB: A better aproximation perhaps using wrap lighting could be implemented here
+            dAttenD = getLightDiffuse(litArgs_worldNormal, dViewDirW, dLightDirW, dLightDirNormW);
+        #else
+            // 16.0 is a constant that is in getFalloffInvSquared()
+            #if LIGHT{i}SHAPE == RECT
+                dAttenD = getRectLightDiffuse(litArgs_worldNormal, dViewDirW, dLightDirW, dLightDirNormW) * 16.0;
+            #elif LIGHT{i}SHAPE == DISK
+                dAttenD = get$DiskLightDiffuse(litArgs_worldNormal, dViewDirW, dLightDirW, dLightDirNormW) * 16.0;
```

Comment:
Wait, what's the `$` for here?

Reply:
I cannot believe you spotted this!

Reply:
Does this finally prove I don't just smash that Approve button? 😉